crates/bevy_app/src/app_builder.rs

```diff
@@ -177,39 +175,6 @@ impl AppBuilder {
         self
     }
 
-    pub fn on_state_enter<T: Clone + Component>(
-        &mut self,
-        stage: impl StageLabel,
-        state: T,
-        system: impl Into<SystemDescriptor>,
-    ) -> &mut Self {
-        self.stage(stage, |stage: &mut StateStage<T>| {
-            stage.on_state_enter(state, system)
-        })
-    }
-
-    pub fn on_state_update<T: Clone + Component>(
-        &mut self,
-        stage: impl StageLabel,
-        state: T,
-        system: impl Into<SystemDescriptor>,
-    ) -> &mut Self {
-        self.stage(stage, |stage: &mut StateStage<T>| {
-            stage.on_state_update(state, system)
-        })
-    }
-
-    pub fn on_state_exit<T: Clone + Component>(
-        &mut self,
-        stage: impl StageLabel,
-        state: T,
-        system: impl Into<SystemDescriptor>,
-    ) -> &mut Self {
-        self.stage(stage, |stage: &mut StateStage<T>| {
-            stage.on_state_exit(state, system)
-        })
-    }
```

Comment:
It may be nice to keep these for convenience.

Reply:
That must've been accidentally removed, I'll add it back

Reply:
Or wait, no. These are missing because right now, there isn't a way to address a system set once it is added, and creating one system set per system isn't the best idea.

Reply:
Restore their usage internally and in examples as well.

Reply:
> creating one system set per system isn't the best idea

Yet that's what's going on in the examples, more or less. It's not the best idea, but it's the best we can do right now. One step at a time...

Reply:
> Yet that's what's going on in the examples
This is because the examples are very small. In more complicated examples, where there is more than one system per set, this pattern reduces code duplication and is overall cleaner. As a compromise, what about something like this?
```rust
impl State<T> {
  fn on_update_set(state: T) -> SystemSet {/**/}
}
``

Reply:
That could work, gonna have to see how example code will look after that.

Alternatively, add `SystemSetLabel`s. It shouldn't be that big a change, and is a decent solution for now - it doesn't even have to be exposed in the API.

The sets' labels could be constructed from the state values themselves, e.g. `on_state_update(MyStage, MyState::MainMenu, my_system.system())` will put the system in a set labelled `(MyState::MainMenu, StateCallback::OnUpdate)`. This won't magically interoperate with user-defined sets, but I think that's fine for now.

Reply:
I'd rather not add `SystemSetLabel`s as part of this PR. If the API is still not usable with something like `on_update_set`, I'd prefer to leave fixing it to a later PR.